### PR TITLE
Flashbar stacked items feature beta test

### PR DESCRIPTION
**Pull request by DaemonCahill**
This PR adds the ability for teams to beta test the notification stacking feature by enabling the option for users to toggle the behavior on and off. There are no changes to the interface in order to prevent widespread awareness and adoption before the feature is truly ready for production after usability testing.

### CODE REVIEW

File: src/flashbar/index.tsx [hunk 1]
```diff
@@ -23,62 +25,142 @@ export default function Flashbar({ items, ...restProps }: FlashbarProps) {
...
+    }
+
     return (
-      <>
-        <TransitionGroup component={null}>
-          {items &&
-            items.map((item, index) => (
-              <Transition
-                transitionChangeDelay={{ entering: TIMEOUT_FOR_ENTERING_ANIMATION }}
+      <div className={clsx(styles['stacking-option'])}>
```
Comment (abdhalees) [inline src/flashbar/index.tsx:56]: I don't see the toggle when opening `http://localhost:8080/#/light/flashbar/stacking/` , how do I activate it?

Comment (DaemonCahill) [reply src/flashbar/index.tsx]: @abdhalees I just re-requested review. There have been significant changes from the first revision. For testing you can use any demo page that has notifications and turn on the secret API by adding `{...{ enableStackingOption: true }}` to the Flashbar component. Some good pages to test with are in the `AppLayout` and `Flashbar` sections.

File: src/flashbar/index.tsx [hunk 2]
```diff
@@ -1,85 +1,177 @@
...
-        </TransitionGroup>
-      </>
+          </div>
+        )}
+
+        {isFlashbarStackExpanded && (
+          <div className={styles.expanded}>{items.map((item, index) => renderItem(item, item.id ?? index))}</div>
+        )}
+
+        <button
```
Comment (abdhalees) [inline src/flashbar/index.tsx:90]: does this button need an aria label?

**Review by abdhalees — CHANGES_REQUESTED** [review]

File: src/flashbar/index.tsx (hunk 2 above)
Comment (DaemonCahill) [reply src/flashbar/index.tsx:90]: Yes, it does. Great catch.

**Review by abdhalees — APPROVED** [review]

File: src/flashbar/index.tsx [hunk 3]
```diff
@@ -1,85 +1,178 @@
...
-  // Motion will be disabled if any of the provided flash messages does not contain an `id`
-
-  const renderFlashItem = (item: FlashbarProps.MessageDefinition, index: number) => (
-    <Flash
-      key={item.id ?? index}
-      // eslint-disable-next-line react/forbid-component-props
-      className={clsx(isRefresh ? styles['flash-refresh'] : '')}
-      {...item}
-    />
+  useEffect(
```
Comment (just-boris) [inline src/flashbar/index.tsx:48]: This should not be an effect: https://beta.reactjs.org/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state

File: src/flashbar/index.tsx [hunk 4]
```diff
@@ -1,85 +1,178 @@
...
+   * flash messages does not contain an `id`.
+   */
+  const motionDisabled =
+    useReducedMotion(ref as any) || !isVisualRefresh || (items && !items.every(item => 'id' in item));
+
+  /**
+   * The `enableStackingOption` property is a hidden boolean that allows for teams
+   * to beta test the flashbar stacking feature.
+   */
+  const enableStackingOption = (restProps as any).enableStackingOption;
```
Comment (just-boris) [inline src/flashbar/index.tsx:44]: Even if it is a beta, better to come up with a solid future-proof name for this property. I think, it should be `stackItems`, but open for other suggestions too.

File: src/flashbar/index.tsx [hunk 5]
```diff
@@ -1,85 +1,178 @@
...
-      </>
+          </div>
+        )}
+
+        {isFlashbarStackExpanded && (
+          <div className={styles.expanded}>{items.map((item, index) => renderItem(item, item.id ?? index))}</div>
+        )}
+
+        <button
+          aria-label={!isFlashbarStackExpanded ? 'Expand stacked notifications' : 'Collapse stacked notifications'}
```
Comment (just-boris) [inline src/flashbar/index.tsx:91]: Let's also expose new properties for this label right away? Even for beta releases, they have to be translatable

File: src/flashbar/index.tsx [hunk 6]
```diff
@@ -1,85 +1,178 @@
...
+    );
+  }
+
   return (
     <div
       {...baseProps}
       className={clsx(baseProps.className, styles.flashbar, styles[`breakpoint-${breakpoint}`])}
       ref={mergedRef}
     >
-      <VisualContext contextName="flashbar">
```
Comment (just-boris) [inline src/flashbar/index.tsx:74]: Why was this changed?

File: src/flashbar/index.tsx (hunk 5 above)
Comment (DaemonCahill) [reply src/flashbar/index.tsx:91]: @just-boris Is there an example somewhere in the code I should follow?

File: src/flashbar/index.tsx (hunk 6 above)
Comment (DaemonCahill) [reply src/flashbar/index.tsx:74]: The visual context function was inserting a `div` around the entire flashbar which applied to the toggle button as well. When I moved that `div` to wrap each flash individually it broke some existing style code relating to margins between flash items. Since the `div` does nothing but apply the visual context class, I added a new function to apply the appropriate class to each flash without the `div`.

File: src/flashbar/index.tsx (hunk 4 above)
Comment (DaemonCahill) [reply src/flashbar/index.tsx:44]: An important distinction that this property only turns on the option to stack for users, but the actual piece of state that controls whether or not the notifications are stacked is separate. I know @connorlanigan had some feedback on the name as well. I will start a Slack thread to see if we can achieve a quick consensus.

File: src/flashbar/index.tsx (hunk 5 above)
Comment (just-boris) [reply src/flashbar/index.tsx:91]: I think it should be `i18nStrings` property, similar to many other our components

File: src/flashbar/index.tsx (hunk 6 above)
Comment (just-boris) [reply src/flashbar/index.tsx:74]: not a big fan of removing the component here (because it will complicate future changes if visual contexts start having additional features), but fine to keep the change for now.

Comment (DaemonCahill) [reply src/flashbar/index.tsx:74]: On this same point I (mistakenly) applied the visual context header class manually in the refactored app layout and content layout components. Inserting a div will absolutely break those components as well, so I think we will want to keep the div-less option. We can discuss in more detail if you want.

Comment (DaemonCahill) [reply src/flashbar/index.tsx:74]: Okay! How would that work when applying a visual context class to an existing element, though?

**Review by just-boris — APPROVED** [review]
Looks good, thanks!
